### PR TITLE
Samsp/prometheus

### PR DESCRIPTION
**Pull request by samsp-msft**
Prometheus sample, showing using metrics classes from Yarp.ReverseProxy.Telemetry.Consumption and prometheus-net to report general and request specific metrics

### CODE REVIEW

**Review by alnikola — COMMENTED** [review]
Looks good overall besides one comment.

File: samples/Prometheus/HttpLoadApp/Program.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace HttpLoadApp
+{
+    class Program
+    {
+        static Random rnd = new Random();
```
Comment (MihaZupan) [inline samples/Prometheus/HttpLoadApp/Program.cs:9]: Random isn't thread-safe

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+
+        public async Task InvokeAsync(HttpContext context)
+        {
+            var startTime = DateTime.Now;
+
+            await _next(context);
+
+            var proxyFeature = context.Features[typeof(IReverseProxyFeature)] as IReverseProxyFeature; 
+            if (proxyFeature != null)
```
Comment (MihaZupan) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs:91]:
```c#
var proxyFeature = context.GetRequiredProxyFeature();
```

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+        public PerRequestYarpMetricCollectionMiddleware(RequestDelegate next)
+        {
+            _next = next;
+        }
+
+
+        public async Task InvokeAsync(HttpContext context)
+        {
+            var startTime = DateTime.Now;
```
Comment (MihaZupan) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs:86]:
```suggestion
            var startTime = DateTime.UtcNow;
```

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+            await _next(context);
+
+            var proxyFeature = context.Features[typeof(IReverseProxyFeature)] as IReverseProxyFeature; 
+            if (proxyFeature != null)
+            {
+                string[] labelvalues = { proxyFeature.RouteSnapshot.ProxyRoute.RouteId, proxyFeature.ClusterSnapshot.Options.Id, proxyFeature.ProxiedDestination.Config.Options.Address };
+                _requestDuration.WithLabels(labelvalues).Observe((startTime - DateTime.Now).TotalMilliseconds);
```
Comment (MihaZupan) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs:94]:
```suggestion
                _requestDuration.WithLabels(labelvalues).Observe((startTime - DateTime.UtcNow).TotalMilliseconds);
```

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                _requestsProcessed.WithLabels(labelvalues).Inc();
+                if (context.Request.ContentLength.HasValue) { _requestContentBytes.WithLabels(labelvalues).Observe(context.Request.ContentLength.Value); }
+                if (context.Response.ContentLength.HasValue) { _responseContentBytes.WithLabels(labelvalues).Observe(context.Response.ContentLength.Value); }
+
+                if (context.Response.StatusCode >= 200 && context.Response.StatusCode < 300) { _requestsSuccessfull.Inc(); }
+                else if (context.Response.StatusCode >= 400 && context.Response.StatusCode < 500) { _requests_error_4xx.Inc(); }
+                else if (context.Response.StatusCode >= 500 && context.Response.StatusCode < 600) { _requests_error_5xx.Inc(); }
```
Comment (MihaZupan) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs:101]: Do these not need dimensions?

We should probably include these as metrics in the proxy by default as well (EventCounters).

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ProxyTelemetryConsumer.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+// Licensed under the MIT License.
+
+using System;
+using Microsoft.AspNetCore.Http;
+using Yarp.ReverseProxy.Service.Proxy;
+using Yarp.ReverseProxy.Telemetry.Consumption;
+
+namespace Yarp.Sample
+{
+    public sealed class ProxyTelemetryConsumer : IProxyTelemetryConsumer
```
Comment (MihaZupan) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ProxyTelemetryConsumer.cs:11]: This class isn't needed for the prometheus sample right?

File: samples/Prometheus/HttpLoadApp/Program.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+        static void Main(string[] args)
+        {
+            HttpClient[] clients = new HttpClient[10];
+            for (int i=0; i<10; i++) { clients[i] = new HttpClient(); }
```
[the remark this answers is not in the document]
Comment (samsp-msft) [reply samples/Prometheus/HttpLoadApp/Program.cs:14]: I think it was an attempt to force multiple connections to the proxy, and appear as multiple clients - but is probably not needed. I'm a bit fuzzy as I wrote that on Friday after a cocktail, so I am not exactly sure what I was thinking at the time.

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs (hunk 2 above)
Comment (samsp-msft) [reply samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs:91]: that throws if the proxy feature isn't present. The proxy isn't there for metrics, and so this fails gracefully.

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ProxyTelemetryConsumer.cs (hunk 6 above)
Comment (samsp-msft) [reply samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ProxyTelemetryConsumer.cs:11]: You are right, I'll remove it

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs (hunk 2 above)
Comment (MihaZupan) [reply samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs:91]: In that case, `context.Features[typeof(IReverseProxyFeature)] as IReverseProxyFeature` can be written as `context.Features.Get<IReverseProxyFeature>()`

File: reverse-proxy.sln [hunk 8]
```diff
@@ -92,6 +92,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ReverseProxy.ConfigFilter.S
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ReverseProxy.Config.Sample", "samples\ReverseProxy.Config.Sample\ReverseProxy.Config.Sample.csproj", "{A19FDE35-C2BB-4326-BA29-C77975FAD4A9}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ReverseProxy.Metrics-Promethius.Sample", "samples\Prometheus\ReverseProxy.Metrics-Promethius.Sample\ReverseProxy.Metrics-Promethius.Sample.csproj", "{31D7E62F-532B-4FAF-87A9-C1BB63AF222A}"
```
Comment (Tratcher) [inline reverse-proxy.sln:95]: Consistent naming:
```suggestion
Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ReverseProxy.Metrics-Promethius.Sample", "samples\Prometheus\ReverseProxy.Metrics.Promethius.Sample\ReverseProxy.Metrics.Promethius.Sample.csproj", "{31D7E62F-532B-4FAF-87A9-C1BB63AF222A}"
```

File: samples/Prometheus/PrometheusSample.sln [hunk 9]
```diff
@@ -0,0 +1,31 @@
+﻿
```
Comment (Tratcher) [inline samples/Prometheus/PrometheusSample.sln:1]: Why a second SLN?

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            var proxyFeature = context.Features.Get<IReverseProxyFeature>();
+            if (proxyFeature != null)
+            {
+                string[] labelvalues = { proxyFeature.RouteSnapshot.ProxyRoute.RouteId, proxyFeature.ClusterSnapshot.Options.Id, proxyFeature.ProxiedDestination.Config.Options.Address };
+                _requestDuration.WithLabels(labelvalues).Observe((startTime - DateTime.UtcNow).TotalMilliseconds);
+                _requestsProcessed.WithLabels(labelvalues).Inc();
+                if (context.Request.ContentLength.HasValue) { _requestContentBytes.WithLabels(labelvalues).Observe(context.Request.ContentLength.Value); }
+                if (context.Response.ContentLength.HasValue) { _responseContentBytes.WithLabels(labelvalues).Observe(context.Response.ContentLength.Value); }
+
+                if (context.Response.StatusCode >= 200 && context.Response.StatusCode < 300) { _requestsSuccessfull.WithLabels(labelvalues).Inc(); }
```
Comment (Tratcher) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs:99]: Also 101 WebSocket Upgrades?

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ReverseProxy.Metrics-Promethius.Sample.csproj [hunk 11]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFrameworks>net5.0;netcoreapp3.1</TargetFrameworks>
+    <OutputType>Exe</OutputType>
+    <RootNamespace>Yarp.Sample</RootNamespace>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="prometheus-net" Version="4.1.1" />
```
Comment (Tratcher) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ReverseProxy.Metrics-Promethius.Sample.csproj:10]: prometheus-net isn't available on our package feeds, we'll have to talk about how to add that. Let's take that offline.

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ReverseProxy.Metrics-Promethius.Sample.sln [hunk 12]
```diff
@@ -0,0 +1,25 @@
+﻿
```
Comment (Tratcher) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ReverseProxy.Metrics-Promethius.Sample.sln:1]: 3 SLNs??

**Review by Tratcher — COMMENTED** [review]
Sorry, no comments on the interesting parts of this, Miha has you covered there. I only have infrastructure questions.

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs (hunk 10 above)
Comment (samsp-msft) [reply samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PerRequestYarpMetricCollectionMiddleware.cs:99]: Ohh.
The bigger problem @Tratcher is that context.Response is returning a default object without useful data on it...

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ReverseProxy.Metrics-Promethius.Sample.sln (hunk 12 above)
Comment (samsp-msft) [reply samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ReverseProxy.Metrics-Promethius.Sample.sln:1]: damn vs

File: samples/Prometheus/HttpLoadApp/Program.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+                {
+                     CreateLoad("http://localhost:5000/" + planet);
+                });
+            }
+           var c = Console.ReadKey();
+        }
+
+        private static async void CreateLoad(string UrlPrefix)
+        {
+            var rnd = new Random();
```
Comment (MihaZupan) [inline samples/Prometheus/HttpLoadApp/Program.cs:26]:
```suggestion
```

Unused now that there's only 1 client

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PrometheusServiceExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            services.AddProxyTelemetryListener();
+            services.AddSingleton<IProxyMetricsConsumer, PrometheusProxyMetrics>();
+            return services;
+        }
+
+#if NET5_0_OR_GREATER
+        public static IServiceCollection AddPrometheusDnsMetrics(this IServiceCollection services)
+        {
+            services.AddSocketsTelemetryListener();
```
Comment (MihaZupan) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/PrometheusServiceExtensions.cs:26]:
```suggestion
            services.AddNameResolutionTelemetryListener();
```

The risk of such typos will go away with #915 where only `AddTelemetryListeners` will be a thing

**Comment by samsp-msft** [comment]
Poking to cause builds to rerun

File: samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ReverseProxy.Metrics.Promethius.Sample.csproj [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+  <PropertyGroup>
+    <TargetFrameworks>net5.0</TargetFrameworks>
+    <OutputType>Exe</OutputType>
+    <RootNamespace>Yarp.Sample</RootNamespace>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="prometheus-net" Version="4.1.1" />
+    <PackageReference Include="prometheus-net.AspNetCore" Version="4.1.1" />
+    <ProjectReference Include="..\..\..\src\ReverseProxy.TelemetryConsumption\Yarp.ReverseProxy.Telemetry.Consumption.csproj" />
```
Comment (Tratcher) [inline samples/Prometheus/ReverseProxy.Metrics-Promethius.Sample/ReverseProxy.Metrics.Promethius.Sample.csproj:12]: Package reference?

**Review by Tratcher — APPROVED** [review]

**Pull request merged**